File: Programmers/cpp/LEVEL_3_이중우선순위큐.cpp
#include <string>
#include <vector>
#include <sstream>
#include <queue>

using namespace std;

vector<int> solution(vector<string> operations) {
	vector<int> answer;
	priority_queue<int> hi;
	priority_queue<int, vector<int>, greater<>> lo;

	for (int i = 0; i < operations.size(); i++) {
		stringstream str(operations[i]);
		char op;
		int num;

		str >> op >> num;

		if (op == 'I') {
			hi.push(num);
			continue;
		}

		if (op == 'D') {
			if (num == 1) {
				if (!hi.empty()) {
					hi.pop();
				}

				continue;
			}

			if (num == -1) {
				while (!hi.empty()) {
					lo.push(hi.top());
					hi.pop();
				}

				if (lo.empty()) {
					continue;
				}

				lo.pop();

				while (!lo.empty()) {
					hi.push(lo.top());
					lo.pop();
				}
			}
		}
	}

	if (hi.empty()) {
		answer.push_back(0);
		answer.push_back(0);
		return answer;
	}

	answer.push_back(hi.top());

	while (!hi.empty()) {
		lo.push(hi.top());
		hi.pop();
	}

	answer.push_back(lo.top());

	return answer;
}

File: Programmers/java/LEVEL_3_이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = {0, 0};
        PriorityQueue<Integer> hi = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> lo = new PriorityQueue<>();

        for (int i = 0; i < operations.length; i++) {
            String[] operation = operations[i].split(" ");
            String op = operation[0];
            int num = Integer.parseInt(operation[1]);

            if (op.equals("I")) {
                hi.add(num);
                continue;
            }

            if (op.equals("D")) {
                if (num == 1) {
                    if (!hi.isEmpty()) {
                        hi.poll();
                    }

                    continue;
                }

                if (num == -1) {
                    while (!hi.isEmpty()) {
                        lo.add(hi.poll());
                    }

                    if (lo.isEmpty()) {
                        continue;
                    }

                    lo.poll();

                    while (!lo.isEmpty()) {
                        hi.add(lo.poll());
                    }
                }
            }
        }

        if (hi.isEmpty()) {
            return answer;
        }

        answer[0] = hi.peek();

        while (!hi.isEmpty()) {
            lo.add(hi.poll());
        }

        answer[1] = lo.peek();

        return answer;
    }
}